src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs

```diff
@@ -3321,6 +3321,94 @@ int IBinaryInteger<BigInteger>.GetByteCount()
             return bits.Length * 4;
         }
 
+        /// <inheritdoc cref="IBinaryInteger{TSelf}.TryWriteBigEndian(Span{byte}, out int)" />
+        bool IBinaryInteger<BigInteger>.TryWriteBigEndian(Span<byte> destination, out int bytesWritten)
+        {
+            AssertValid();
+            uint[]? bits = _bits;
+
+            int byteCount = (bits is null) ? sizeof(int) : bits.Length * 4;
+
+            if (destination.Length >= byteCount)
+            {
+                if (bits is null)
+                {
+                    int value = BitConverter.IsLittleEndian ? BinaryPrimitives.ReverseEndianness(_sign) : _sign;
+                    Unsafe.WriteUnaligned(ref MemoryMarshal.GetReference(destination), value);
+                }
+                else if (_sign >= 0)
+                {
+                    // When the value is positive, we simply need to copy all bits as little endian
+
+                    ref byte address = ref MemoryMarshal.GetReference(destination);
+                    address = ref Unsafe.Add(ref address, (bits.Length - 1) * sizeof(uint));
+
+                    for (int i = 0; i < bits.Length; i++)
+                    {
+                        uint part = bits[i];
+
+                        if (BitConverter.IsLittleEndian)
+                        {
+                            part = BinaryPrimitives.ReverseEndianness(part);
+                        }
+
+                        Unsafe.WriteUnaligned(ref address, part);
+                        address = ref Unsafe.Subtract(ref address, sizeof(uint));
+                    }
+                }
+                else
+                {
+                    // When the value is negative, we need to copy the two's complement representation
+                    // We'll do this "inline" to avoid needing to unnecessarily allocate.
+
+                    ref byte address = ref MemoryMarshal.GetReference(destination);
+                    address = ref Unsafe.Add(ref address, (bits.Length - 1) * sizeof(uint));
+
+                    int i = 0;
+                    uint part;
+
+                    do
+                    {
+                        part = ~bits[i] + 1;
+
+                        if (BitConverter.IsLittleEndian)
+                        {
+                            part = BinaryPrimitives.ReverseEndianness(part);
+                        }
+
+                        Unsafe.WriteUnaligned(ref address, part);
+                        address = ref Unsafe.Subtract(ref address, sizeof(uint));
+
+                        i++;
+                    }
+                    while ((part == 0) && (i < bits.Length));
```

Comment:
What spins here, and do we have boundary tests for it?  (e.g. something does the do and exits to the other loop, something else runs the body twice, something runs thrice, and maybe 4 isn't possible?)

(A comment in the code to help future maintainers would be helpful).

Reply:
No sure what you're asking for here? This handles the conversion from `one's complement` to `two's complement`.

The first loop handles `0x0000_0000` as the one's complement of that is `0x0000_0000`. The second loop handles the remaining bits in the value.
* `0x0000_0000_0000_0000_0000_0000` becomes `0x0000_0000_0000_0000_0000_0000`
* `0x0000_0000_0000_0000_0000_0001` becomes `0xFFFF_FFFF_FFFF_FFFF_FFFF_FFFF`
* `0x0000_0000_0000_0001_0000_0000` becomes `0x0000_0000_FFFF_FFFF_0000_0000`

This handles what is logically `~x + 1`

Reply:
I'm trying to understand what values of a BigInteger might cause the do/while to run more than once.  Given that there's all of 1s complement, 2s complement, little endian, and big endian present, I can't follow the data flow in my head.

* Negative values `-1` .. `int.MinValue + 1` all fit in `_sign` and `_bits` is null, so none of those are relevant.
* `int.MinValue` produces a one element `_bits` value, so doesn't loop.
* `int.MinValue - 1` (0xFFFFFFFF_7FFFFFFF) is, I believe, stored as `{ 0x00000000, 0x80000000 }`
  * The first time through the loop we load the 0.  So now do ~0 + 1, which I guess goes back to 0.
    * Yay, this value looped.
  * The second time through we load the 0x8000_0000 and produce, uh, 0x8000_0000?
    * It's not zero, so we'd exit, but we're also out of data, so we'd exit anyways.

So perhaps the answer is "anything in the tests ending in multiples of 4 zero-byte values (BE) went through the top loop (trailing_zero_bytes_count / 4) + 1 times".

If that's the case, I'd expect to see tests that
* Don't end in 4 zeros, but are at least 8 bytes (do+break+second_loop)
* End in 4 zeros and are 8 bytes (do+do+return)
* End in 4 zeros and are more than 8 bytes (do+do+break+second_loop)
* End in 8 zeros and are 12 bytes (do+do+do+return)
* End in 8 zeros and are more than 12 bytes (do+do+do+break+second_loop)

Ideally with each of the second_loop cases having a one/more_than_one case (and super ideally as one/two/more_than_two, since sometimes "two" is special)

If I understand the states now, I don't see any tests that produce 12 or more bytes, so there's nothing that runs that top loop 3 or more times.  

---

The thing I'd hope for here is a comment talking about some of the boundary cases.  i.e. a particular value that will run the top loop twice, then the bottom loop twice, but one more (or less?) would have a different execution characteristic (which I'm not feeling knowledgeable enough to predict if that means 1/3 or 3/1, or a black hole, or what).  And those same values being codified in tests.

Reply:
For this method in particular, we have tests covering everything but the `12 + byte` case because it requires significant more setup to do and represent.

I can certainly add it, but its not interesting IMO. This logic is the same as we have in `WriteLittleEndian` and that is used in `DangerousMakeTwosComplement`, etc. The only difference is that it writes from the end of the buffer to the beginning, rather than from the beginning to the end.

Reply:
Fixed. There was indeed an issue with a specific edge case value due to it not accounting for needing an extra "part" to hold the sign.

In particular, `signed.MinValue - 1` is represented in `one's complement` as `0x8000_0000, 0x0000_0001`, swapped to `two's complement` this is `0x7FFF_FFFF, 0xFFFF_FFFF` which is of course `signed.MaxValue` instead. We need it to be `0xFFFF_FFFF, 0x7FFF_FFFF, 0xFFFF_FFFF` so that the sign is still tracked instead.